drivers: ethernet: stm32: factorize clock handles

etienne-lms opened the pull request:
Factorize STM32 interface clocks configuration in a single array. This change eases later integration of other SoCs with different clocks names while the driver only has to enable (possibly disable) the clocks on a single sequence.

Ethernet MAC clocks are moved from MAC node to ethernet controller (parent) node for sake of simplicity.

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -420,17 +413,18 @@ static void eth0_irq_config(void)
+static const struct stm32_pclken eth0_pclken[] = {
+	/* Ethernet parent node clock is always the first in the array */
+	STM32_CLOCK_INFO_BY_NAME(DT_INST_PARENT(0), stm_eth),
+	LISTIFY(DT_NUM_CLOCKS(DT_DRV_INST(0)), STM32_CLOCK_INFO, (,), DT_DRV_INST(0))
```
How about we move all clocks to the parent?

etienne-lms committed:
drivers: ethernet: stm32: factorize clock handles

Factorize STM32 interface clocks configuration in a single array.
This change eases later integration of other SoCs with different
clocks names while the driver only has to enable (possibly disable)
the clocks on a single sequence.

Suggested-by: Fin Maaß <f.maass@vogl-electronic.com>
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
I've rebased to fix merge conflicts. No change.

etienne-lms committed:
dts: arm: st: move ethernet clocks from mac to controller node

Move the ethernet controller clocks from mac node to controller node.
This change simplifies how clocks are gathered and handled in STM32
ethernet drivers.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: ethernet: stm32: use STM32_DT_INST_CLOCK_INFO_BY_NAME()

Update STM32 ethernet driver to use STM32_DT_INST_CLOCK_INFO_BY_NAME()
helper macro to always get all clock information, not only the bus ID
and bit position.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: ethernet: stm32: prepare move of MAC clocks to parent node

Prepare move of STM23 ethernet MAC clocks to the controller node (parent
node). For sake of simplicity, define a filed for all possible clocks in
struct eth_stm32_hal_dev_cfg, a later change will replace the whole
with a single STM32 clock instance (struct stm32_pclken) array pointer.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: ethernet: stm32: finalize move of clocks to controller node

Remove helper macro used to transition from MAC clocks defined
by the MAC node to definition in the controller (parent) node.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -104,17 +104,11 @@ extern ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB];
+	const struct stm32_pclken *pclken;
+	size_t pclken_cnt;
+#ifdef CONFIG_PTP_CLOCK_STM32_HAL
+	/* Index of the clock that give the ethernet clock rate */
```
```suggestion
	/* Index of the clock that gives the ethernet clock rate */

mathieuchopstm commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -152,30 +152,14 @@ static int eth_initialize(const struct device *dev)
-#if DT_INST_CLOCKS_HAS_NAME(0, eth_ker)
-	ret |= clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
-				       (clock_control_subsys_t)&cfg->pclken_ker,
-				       NULL);
```
This is straight up gone?! Or am I missing something?

etienne-lms replied:
Addressed.

etienne-lms commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -152,30 +153,21 @@ static int eth_initialize(const struct device *dev)
-#endif
+	/* Enable clocks */
+	for (size_t n = 0; n < cfg->pclken_cnt; n++) {
+		if (n == cfg->kclk_sel_idx) {
```
`cfg->kclk_sel_idx` could be removed by testing here:
```suggestion
		if (pclken[n].bus < STM32_PERIPH_BUS_MIN) {
```
Maybe less explicit and fragile.
